addons/account/models/partner.py

```diff
@@ -172,52 +172,34 @@ def write(self, vals):
                 vals['zip_from'], vals['zip_to'] = self._convert_zip_values(zip_from or rec.zip_from, zip_to or rec.zip_to)
         return super(AccountFiscalPosition, self).write(vals)
 
-    @api.model
-    def _get_fpos_by_region(self, country_id=False, state_id=False, zipcode=False, vat_required=False):
-        if not country_id:
-            return False
-        base_domain = self._prepare_fpos_base_domain(vat_required)
-        null_state_dom = state_domain = [('state_ids', '=', False)]
-        null_zip_dom = zip_domain = [('zip_from', '=', False), ('zip_to', '=', False)]
-        null_country_dom = [('country_id', '=', False), ('country_group_id', '=', False)]
-
-        if zipcode:
-            zip_domain = [('zip_from', '<=', zipcode), ('zip_to', '>=', zipcode)]
-
-        if state_id:
-            state_domain = [('state_ids', '=', state_id)]
-
-        domain_country = base_domain + [('country_id', '=', country_id)]
-        domain_group = base_domain + [('country_group_id.country_ids', '=', country_id)]
-
-        # Build domain to search records with exact matching criteria
-        fpos = self.search(domain_country + state_domain + zip_domain, limit=1)
-        # return records that fit the most the criteria, and fallback on less specific fiscal positions if any can be found
-        if not fpos and state_id:
-            fpos = self.search(domain_country + null_state_dom + zip_domain, limit=1)
-        if not fpos and zipcode:
-            fpos = self.search(domain_country + state_domain + null_zip_dom, limit=1)
-        if not fpos and state_id and zipcode:
-            fpos = self.search(domain_country + null_state_dom + null_zip_dom, limit=1)
-
-        # fallback: country group with no state/zip range
-        if not fpos:
-            fpos = self.search(domain_group + null_state_dom + null_zip_dom, limit=1)
-
-        if not fpos:
-            # Fallback on catchall (no country, no group)
-            fpos = self.search(base_domain + null_country_dom, limit=1)
-        return fpos
-
     def _get_vat_valid(self, delivery, company=None):
         """ Hook for determining VAT validity with more complex VAT requirements """
         return bool(delivery.vat)
 
-    def _prepare_fpos_base_domain(self, vat_required):
+    def _get_fpos_ranking_functions(self, partner):
         return [
-            *self._check_company_domain(self.env.company),
-            ('auto_apply', '=', True),
-            ('vat_required', '=', vat_required),
+            ('vat_required', lambda fpos: (
+                not fpos.vat_required
+                or (self._get_vat_valid(partner, self.env.company) and 2)
```

Comment:
Something is off: This is only 0 when the vat is required on the fpos and there is no vat, while one or the other way, the logic was changed before and vat_required has to be equal to vat_required.  I could say: put it back  in the base domain, but this might work as well:  (just to follow the same logic in a very special case where zip would be set/not set e.g.)
```suggestion
                or (self._get_vat_valid(partner, self.env.company) and 200)
```

But this does not work when there is no vat on the partner, because then it should never get into the vat required ones.  So we need (not fpos.required and not self._get_vat_valid(partner, company) and 1 or 0)  (but then the terms need to be juggled)

So we need something like:
self._get_vat_valid(partner, self.env.company) and 200 or 1 if fpos.vat_required else not self._get_vat_valid(partner, company)

Reply:
I didn't get it either @jco-odoo :/

Reply:
Do you know how tuple comparison works? :smile:

Reply:
So, tuple comparison, no sum.